### DISCUSSION TITLE
Possible to initialize an FlxState with different data for each level?

### DISCUSSION THREAD

**Discussion by ChillyCider** [HaxeFlixel/flixel#2477] [Q&A] [answered]
Hi! I'm making a little puzzle game to exercise myself with HaxeFlixel. Please forgive what may be a very basic question. I have looked through the docs but I haven't found any hints there.

Is there a recommended way to pass extra data to an FlxState when I'm switching to it? If so, how could I do this? The HaxeFlixel docs warn against overriding the FlxState constructor, so I've purposefully avoided doing that.

The reason I want to do this--I have my puzzles defined as JSON resources, and I'm hoping to be able to load up a specific puzzle into the PuzzleState as needed.

Thanks for your time!

**Comment by RichardBray** [comment, marked as answer]
I guess the answer to this question depends on how you're loading up your FlxStates.

1. If you're using `FlxG.switchState`. You can pass arguments into your level like this:
```haxe
  FlxG.switchState(new LevelOne(argument1, argument2));
```
Then you can load the arguments into your class constructor.
Happy to explain more if it doesn't make sense.

2. If you have an array of states then check out this thread:
https://github.com/HaxeFlixel/flixel/discussions/2400

**Reply by ChillyCider** [reply]
> 1. If you're using `FlxG.switchState`. You can pass arguments into your level like this:
> 
> ```hx
>   FlxG.switchState(new LevelOne(argument1, argument2));
> ```
> 
> Then you can load the arguments into your class constructor. Happy to explain more if it doesn't make sense.

Thanks for your assistance.

Is it okay to do it like that? The FlxState docs say: "We do NOT recommend overriding the constructor, unless you want some crazy unpredictable things to happen!"

**Reply by Gama11** [reply]
The issue is mostly that `FlxG.resetState()` breaks.

**Reply by Geokureli** [reply]
also it's a very bad idea to put any real init logic inside the constructor, since the constructor is usually called before state-switching has started. and you're initial state can't have constructor args (optional args might work). But I use FlxState constructors all the time to send data to the new state. it's cleaner than using global vars

[Here's my attempt](https://github.com/HaxeFlixel/flixel/pull/2479) to clear up the doc language on constructors

**Reply by ChillyCider** [reply]
Thanks for the help everybody! It sounds like constructor args are the way to go. I don't need `FlxG.resetState()` so this works out great.
